### PR TITLE
ci/cirrus: use vagrant from hashicorp repo, bump Fedora to 38, bump bats

### PR DESCRIPTION
**Pull request by kolyshkin**
### 1. Use vagrant from hashicorp repo.

A version of vagrant available from the stock repos (2.2.19) is too old and contains a bug that prevents downloading Fedora 38 image. The error looks like this:

```
    default: Downloading: https://vagrantcloud.com/fedora/boxes/38-cloud-base/versions/38.20230413.1/providers/libvirt.box

An error occurred while downloading the remote file. The error
message, if any, is reproduced below. Please fix this error and try
again.

HTTP/2 stream 0 was not closed cleanly: PROTOCOL_ERROR (err 1)
```

Apparently, this is caused by https://github.com/hashicorp/vagrant/issues/12921 (plus a combination of some server-side changes, I guess).

The code in question was added 5 years ago in vagrant 2.1.3 (https://github.com/hashicorp/vagrant/commit/447e58dac3edc77f655379729d172a3503d7019c) , and it was fixed (by https://github.com/hashicorp/vagrant/commit/c7ef689d86294bcdb8ae5d31ddabcba416e3382d) in vagrant 2.3.1.

As Ubuntu 22.04 comes with vagrant 2.2.19, we hit the issue.

The solution is to use packages from hashicorp repo, which currently has vagrant 2.3.4.
This resolves the problem of downloading the latest Fedora image.
    
Also, vagrant-libvirt plugin from Ubuntu repos is not working with
vagrant from hashicorp, so switch to using "vagrant plugin install".
The downside it, this takes extra 4 minutes or so in our CI, and I
am not sure how to cache it or speed it up.

### 3. ci: bump bats 1.8.2 -> 1.9.0
    
As Fedora 38 uses bats 1.9.0, let's switch to this version in other
places.

### CODE REVIEW

File: .cirrus.yml [hunk 1]
```diff
@@ -37,6 +37,8 @@ task:
     # -----
     cat /proc/cpuinfo
   install_libvirt_vagrant_script: |
+    wget -O- https://apt.releases.hashicorp.com/gpg | sudo apt-key add -
```
Comment (thaJeztah) [inline .cirrus.yml:40]: note that `apt-key` is deprecated (also see https://github.com/docker/docs/pull/11990), so would be something like;

```bash
 wget-O- https://apt.releases.hashicorp.com/gpg | sudo gpg --dearmor -o /etc/apt/keyrings/hashicorp.gpg
```

or (making sure the directory exists, and set permissions; see https://github.com/docker/docs/pull/17070)

```bash
 sudo install -m 0755 -d /etc/apt/keyrings
 wget-O- https://apt.releases.hashicorp.com/gpg | sudo gpg --dearmor -o /etc/apt/keyrings/hashicorp.gpg
 sudo chmod a+r /etc/apt/keyrings/hashicorp.gpg
```

**Comment by kolyshkin** [comment]
Ahh, the next problem is vagrant-libvirt plugin from ubuntu repo is not recognized by vagrant from the hashicorp repo (https://github.com/hashicorp/vagrant/issues/12657).

File: .cirrus.yml (hunk 1 above)
Comment (kolyshkin) [reply .cirrus.yml]: Thanks; done.

File: .cirrus.yml [hunk 2]
```diff
@@ -37,9 +37,15 @@ task:
...
   install_libvirt_vagrant_script: |
+    curl -fsSL https://apt.releases.hashicorp.com/gpg | sudo gpg --dearmor -o /usr/share/keyrings/hashicorp-archive-keyring.gpg
+    echo "deb [signed-by=/usr/share/keyrings/hashicorp-archive-keyring.gpg] https://apt.releases.hashicorp.com $(lsb_release -cs) main" | sudo tee /etc/apt/sources.list.d/hashicorp.list
+    sudo sed -i 's/^# deb-src/deb-src/' /etc/apt/sources.list
     apt-get update
-    apt-get install -y libvirt-daemon libvirt-daemon-system vagrant vagrant-libvirt
+    apt-get install -y libvirt-daemon libvirt-daemon-system vagrant
     systemctl enable --now libvirtd
+    apt build-dep -y vagrant ruby-libvirt
+    apt install -y --no-install-recommends libxslt-dev libxml2-dev libvirt-dev ruby-bundler ruby-dev zlib1g-dev
```
Comment (AkihiroSuda) [inline .cirrus.yml:47]: apt-get

Comment (kolyshkin) [reply .cirrus.yml]: fixed

**Comment by kolyshkin** [comment]
Filed a bug to Ubuntu: https://bugs.launchpad.net/vagrant/+bug/2017828

**Comment by kolyshkin** [comment]
Do we want to backport it? I'm not entirely sure. @AkihiroSuda if you feel like it, just add a label.

**Review by AkihiroSuda — APPROVED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by AkihiroSuda** [comment]
> Filed a bug to Ubuntu: https://bugs.launchpad.net/vagrant/+bug/2017828

How can we get Ubuntu to look at this?

**Pull request merged**